File: src/types/schema.ts
// Column definition
export interface Column {
  name: string;
  dataType: string;
  isNullable: boolean;
  isPrimaryKey: boolean;
  sourceTable?: string; // Source table name (for view columns)
  sourceColumn?: string; // Source column name (for view columns)
}

// Table node representation
export interface TableNode {
  id: string; // Format: "schema.table"
  name: string; // Table name only
  schema: string; // Schema name (e.g., "dbo")
  columns: Column[];
}

// View node representation
export interface ViewNode {
  id: string; // Format: "schema.view"
  name: string; // View name only
  schema: string; // Schema name (e.g., "dbo")
  columns: Column[];
  definition: string; // SQL definition
  referencedTables: string[]; // List of table/view IDs referenced in the view
}

// Foreign key relationship
export interface RelationshipEdge {
  id: string; // Unique FK identifier
  from: string; // Source table ID ("schema.table")
  to: string; // Target table ID ("schema.table")
  fromColumn: string; // FK column in source
  toColumn: string; // Referenced column in target
}

// Trigger definition
export interface Trigger {
  id: string; // Format: "schema.table.trigger_name"
  name: string;
  schema: string;
  tableId: string; // Parent table ID ("schema.table")
  triggerType: string; // e.g., "AFTER", "INSTEAD OF"
  isDisabled: boolean;
  firesOnInsert: boolean;
  firesOnUpdate: boolean;
  firesOnDelete: boolean;
  definition: string; // SQL definition
  referencedTables: string[]; // List of table/view IDs referenced in the trigger
}

// Stored procedure parameter
export interface ProcedureParameter {
  name: string;
  dataType: string;
  isOutput: boolean;
}

// Stored procedure definition
export interface StoredProcedure {
  id: string; // Format: "schema.procedure_name"
  name: string;
  schema: string;
  procedureType: string; // e.g., "SQL_STORED_PROCEDURE"
  parameters: ProcedureParameter[];
  definition: string; // SQL definition
  referencedTables: string[]; // List of table/view IDs referenced in the procedure
}

// Complete schema graph
export interface SchemaGraph {
  tables: TableNode[];
  views: ViewNode[];
  relationships: RelationshipEdge[];
  triggers: Trigger[];
  storedProcedures: StoredProcedure[];
}

// Connection parameters
export interface ConnectionParams {
  server: string;
  database: string;
  username: string;
  password: string;
  trustServerCertificate?: boolean;
}

File: src-tauri/src/types/schema.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct Column {
    pub name: String,
    pub data_type: String,
    pub is_nullable: bool,
    pub is_primary_key: bool,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub source_table: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub source_column: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TableNode {
    pub id: String,
    pub name: String,
    pub schema: String,
    pub columns: Vec<Column>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ViewNode {
    pub id: String,
    pub name: String,
    pub schema: String,
    pub columns: Vec<Column>,
    pub definition: String,
    pub referenced_tables: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RelationshipEdge {
    pub id: String,
    pub from: String,
    pub to: String,
    pub from_column: String,
    pub to_column: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProcedureParameter {
    pub name: String,
    pub data_type: String,
    pub is_output: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Trigger {
    pub id: String,
    pub name: String,
    pub schema: String,
    pub table_id: String,
    pub trigger_type: String,
    pub is_disabled: bool,
    pub fires_on_insert: bool,
    pub fires_on_update: bool,
    pub fires_on_delete: bool,
    pub definition: String,
    pub referenced_tables: Vec<String>,
    pub affected_tables: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StoredProcedure {
    pub id: String,
    pub name: String,
    pub schema: String,
    pub procedure_type: String,
    pub parameters: Vec<ProcedureParameter>,
    pub definition: String,
    pub referenced_tables: Vec<String>,
    pub affected_tables: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ScalarFunction {
    pub id: String,
    pub name: String,
    pub schema: String,
    pub function_type: String,
    pub parameters: Vec<ProcedureParameter>,
    pub return_type: String,
    pub definition: String,
    pub referenced_tables: Vec<String>,
    pub affected_tables: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaGraph {
    pub tables: Vec<TableNode>,
    pub views: Vec<ViewNode>,
    pub relationships: Vec<RelationshipEdge>,
    pub triggers: Vec<Trigger>,
    pub stored_procedures: Vec<StoredProcedure>,
    pub scalar_functions: Vec<ScalarFunction>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum AuthType {
    #[default]
    SqlServer,
    Windows,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ConnectionParams {
    pub server: String,
    pub database: String,
    #[serde(default)]
    pub auth_type: AuthType,
    #[serde(default)]
    pub username: Option<String>,
    #[serde(default)]
    pub password: Option<String>,
    #[serde(default)]
    pub trust_server_certificate: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ServerConnectionParams {
    pub server: String,
    #[serde(default)]
    pub auth_type: AuthType,
    #[serde(default)]
    pub username: Option<String>,
    #[serde(default)]
    pub password: Option<String>,
    #[serde(default)]
    pub trust_server_certificate: bool,
}
